File: 2023/haskell/Day1/Day1.hs
{-# LANGUAGE ImportQualifiedPost #-}

import Data.Char (isDigit)
import Data.List (isPrefixOf)
import Data.Map.Strict qualified as M
import System.IO ()

word2num :: M.Map String String
word2num =
  M.fromList
    [ ("one", "1"),
      ("two", "2"),
      ("three", "3"),
      ("four", "4"),
      ("five", "5"),
      ("six", "6"),
      ("seven", "7"),
      ("eight", "8"),
      ("nine", "9")
    ]

findFirstDigit :: String -> M.Map String String -> Maybe Char
findFirstDigit [] _ = Nothing
findFirstDigit (c : cs) wordMap
  | isDigit c = Just c
  | otherwise = case [num | (word, num) <- M.toList wordMap, word `isPrefixOf` (c : cs)] of
      (n : _) -> Just (head n)
      [] -> findFirstDigit cs wordMap

findLastDigit :: String -> M.Map String String -> Maybe Char
findLastDigit s wordMap = findFirstDigit (reverse s) (M.mapKeys reverse wordMap)

extractCalibrationValue :: String -> M.Map String String -> Int
extractCalibrationValue s wordMap =
  case (findFirstDigit s wordMap, findLastDigit s wordMap) of
    (Just first, Just last) -> read [first, last]
    _ -> 0

processFile :: String -> M.Map String String -> IO Int
processFile [] _ = do
  putStrLn "No filename was provided"
  return 0
processFile filename wordMap = do
  file <- readFile filename
  let fileLines = lines file
  let calibrationValues = map (`extractCalibrationValue` wordMap) fileLines
  return $ sum calibrationValues

main :: IO ()
main = do
  let filePath = "..\\..\\day1_input.txt"
  result <- processFile filePath word2num
  putStrLn $ "The sum of all calibration values is: " ++ show result

File: 2023/python/Day1.py
import re

word2num = {
    'one': '1',
    'two': '2',
    'three': '3',
    'four': '4',
    'five': '5',
    'six': '6',
    'seven': '7',
    'eight': '8',
    'nine': '9',
    'zero': '0'
}

def convert2nums(line, map):
    pattern = '(?=(' + '|'.join(map.keys()) + '|\\d))'
    return re.sub(pattern, lambda x: map.get(x.group(1), x.group(1)), line)

def process_file(filename, map):
    total = 0
    with open(filename, 'r') as f:
        for line in f:
            converted_line = convert2nums(line.strip(), map)
            numbers = re.findall(r'\d', converted_line)
            if numbers:
                num = int(numbers[0] + numbers[-1])
                total += num
    return total

total = process_file("../day1_input.txt", word2num)
print(total)
